### PR TITLE
 Send Less Client Reports When Rate Limited 

### LINKED ISSUE #2370: Send Less Client Reports When Rate Limited
### Description

The Dart SDK attaches the client report before applying the rate limit in the `client`.

https://github.com/getsentry/sentry-dart/blob/0880a97f2be69b0bbd4e9fb8460afd39a73b5033/dart/lib/src/sentry_client.dart#L630-L637

In the `http_trasnport`, it then applies the rate limit
https://github.com/getsentry/sentry-dart/blob/fe6dcac6041d41f2b6ff38f507a75470ea94b2f4/dart/lib/src/transport/http_transport.dart#L37-L42

Because clients can assume that client reports never get rate limited ([develop docs](https://develop.sentry.dev/sdk/telemetry/client-reports/#sdk-side-recommendations)), the Dart SDK continues to send an envelope with a client report even when the SDK is fully rate-limited and should drop all envelope items. Instead, the Dart SDK should only add the client report envelope item if the envelope has items to send.

The downside of this is that when the SDK is fully rate-limited for a long time or the user drops all events in beforeSend, the SDK may never send a client report. This can be addressed by checking in the HTTP transport if the ClientReports has  more than x discarded events (I guess 10 is a good start), and then send an envelope with only the client report. This is related to https://github.com/getsentry/sentry-cocoa/issues/4468.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- [x] Add client reports after rate limit filtering.
- [x] Try to send client reports only after getting fully rate limited 10 times.

## :bulb: Motivation and Context

Closes #2370

## :green_heart: How did you test it?

- Unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/transport/http_transport.dart [hunk 1]
```diff
@@ -35,14 +35,16 @@ class HttpTransport implements Transport {
...
-    if (filteredEnvelope == null) {
-      return SentryId.empty();
-    }
-    filteredEnvelope.header.sentAt = _options.clock();
+    // final filteredEnvelope = _rateLimiter.filter(envelope);
+    // if (filteredEnvelope == null) {
+    //   return SentryId.empty();
+    // }
+    // final clientReport = _options.recorder.flush();
+    // envelope.addClientReport(clientReport);
```
Comment (buenaflor) [inline dart/lib/src/transport/http_transport.dart]: I guess this is leftover and can be removed?

File: dart/lib/src/transport/client_report_transport.dart [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import 'package:meta/meta.dart';
+import '../../sentry.dart';
+import '../client_reports/client_report_recorder.dart';
+import '../sentry_envelope_header.dart';
+import 'rate_limiter.dart';
+
+/// Decorator that handles attaching of client reports in tandem with rate
+/// limiting. The rate limiter is optional.
```
Comment (buenaflor) [inline dart/lib/src/transport/client_report_transport.dart:7]: wdym by it's optional? 

I think it's always added to our transports so it's also applying it to native, no?

Comment (denrase) [reply dart/lib/src/transport/client_report_transport.dart:7]: We don't rate limit on the flutter side when using the file based transport. The native libs have this responsibility. In order to reuse for both transports we support now, the injected rate limiter is optional.

Comment (buenaflor) [reply dart/lib/src/transport/client_report_transport.dart:7]: ah got it

```dart
    if (options.transport is NoOpTransport) {
      rateLimiter = RateLimiter(options);
      options.transport = HttpTransport(options, rateLimiter);
    }
    options.transport = ClientReportTransport(
      rateLimiter,
      options,
      options.transport,
    );
```

was confused at first, to me it looked like at first glance the rateLimiter is always active even for FileSystemTransrpot, didn't realize that it's null there

is it possible to make this a bit more obvious? or maybe add a comment or similar

**Review by buenaflor — APPROVED** [review]

**Pull request merged**